### PR TITLE
Reflects ListStore Grand Refactor from FluxC for post list

### PR DESCRIPTION
**Pull request by oguzkocer**
This is a draft PR that reflects the changes made in `ListStore` for the grand refactor. I am only opening this PR to make it easy to test the changes and to make it clear how the new components are intended to be used. We'll either open a fresh PR or add more details to this one before it's marked ready for review. For now, please see the FluxC PR for details: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1193

---
**Updated PR description:**

We recently finished a major refactor for `ListStore` in FluxC and as part of it we decided to move some of the feature specific logic to where the feature is implemented. This PR addresses these changes.

When I opened this PR as a draft PR, I thought there would be a lot more to explain, but reviewing the changes again, I can't really find much to talk about. It's mostly moving things around to reflect the FluxC changes and nothing else.

The only important issue that's worth mentioning here is that we no longer prevent duplicate fetch requests for posts in the `PostListItemDataSource`. That has been moved to this repo as a `TODO` because we don't know how we'll handle that yet. (time constraints) So, in its current state, if the user visits the posts page for the first time and has a lot of posts, it's very likely that the list will behave weirdly because too many requests will happen at the same time (in most cases for a very small number of posts) and result in a lot of refreshes. This issue will certainly be fixed before this feature is merged to `develop`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                // We are creating a list of local and remote ids, so other type of identifiers don't matter
+                EndListIndicatorIdentifier -> null
+            }
+        }
+
+        val postList = postStore.getPostsByLocalOrRemotePostIds(localOrRemoteIds, listDescriptor.site)
+
+        // Convert the post list into 2 maps with local and remote ids as keys
+        val localPostMap = postList.associateBy { LocalId(it.id) }
+        val remotePostMap = postList.filter { it.remotePostId != 0L }.associateBy { RemoteId(it.remotePostId) }
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Nitpick: Can we replace it.remotePostId != 0L with !it.isLocalDraft or is there any other scenario when the remotePostId == 0?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        val localItems = postStore.getLocalPostIdsForDescriptor(listDescriptor)
+                .map { LocalPostId(id = it) }
+        val remoteItems = remoteItemIds.map { RemotePostId(id = it) }
+        val actualItems = localItems + remoteItems
+
+        val endListItem = if (isListFullyFetched && actualItems.isNotEmpty()) {
+            listOf(EndListIndicatorIdentifier)
+        } else emptyList()
+        val itemsToHide = getItemIdsToHide(listDescriptor)
+        // We only want to show the end list indicator if the list is fully fetched and it's not empty
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Nitpick: Shouldn't this comment be moved 4 lines above?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        val postList = postStore.getPostsByLocalOrRemotePostIds(localOrRemoteIds, listDescriptor.site)
+
+        // Convert the post list into 2 maps with local and remote ids as keys
+        val localPostMap = postList.associateBy { LocalId(it.id) }
+        val remotePostMap = postList.filter { it.remotePostId != 0L }.associateBy { RemoteId(it.remotePostId) }
+
+        // Fetch the missing posts
+        val remoteIdsToFetch: List<RemoteId> = localOrRemoteIds.mapNotNull { it as? RemoteId }
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Nitpick: We may want to consider moving this into its own method. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                .filter { !remotePostMap.containsKey(it) }
+        fetchPosts(listDescriptor.site, remoteIdsToFetch)
+
+        val transformFromNullablePost = { localOrRemoteId: LocalOrRemoteId, post: PostModel? ->
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Nitpick: I feel like this lambda makes the code just more difficult to understand. How do you feel about moving it into a regular method - we could call it something like "transformToUiState"?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemType.kt [hunk 5]
```diff
@@ -1,18 +1,27 @@
...
 import org.wordpress.android.widgets.PostListButtonType
 
-class PostListItemUiState(
-    val data: PostListItemUiStateData,
-    val actions: List<PostListItemAction>,
-    val onSelected: () -> Unit
-)
+sealed class PostListItemType {
+    // TODO: Can we find a better name for this now that it's a PostListItemType?
+    class PostListItemUiState(
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemType.kt]: Maybe something like "ExpandedPost" as a preparation for "CollapsedPost" ... or just LoadedItem/ReadyItem.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt (hunk 1 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: I am not sure if those two properties are connected or not, but if they are in the FluxC side, I think we'll have to assume the implementation detail here. We need the remote id to be able to fetch posts and even if they mean the same thing, I feel like this expresses the intend a bit better.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt (hunk 2 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Fixed in 093558595838c2a8219b6991cb9e1bb4bbc48e62.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt (hunk 3 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Fixed in 093558595838c2a8219b6991cb9e1bb4bbc48e62.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt (hunk 4 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt]: Fixed in 093558595838c2a8219b6991cb9e1bb4bbc48e62.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemType.kt (hunk 5 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemType.kt]: I am not sure if the `ExpandedPost` will have the meaning we'd want in where it's consumed, I'd like to think on this a bit more before we commit to something if it's OK with you.

**Comment by oguzkocer** [comment]
@malinajirka Thank you for the review. I've made most of the changes you mentioned and replied to a couple comments with my thoughts. Let me know what you think!

**Comment by malinajirka** [comment]
Awesome, thanks for the changes @oguzkocer! LGTM

**Pull request merged**